pkg/storage/stores/local/archiver.go

```diff
@@ -0,0 +1,493 @@
+
+	for _, object := range objects {
+		uploader := strings.Split(object.Key, "/")[1]
+		listedUploaders[uploader] = struct{}{}
+
+		// Checking whether file was updated in the store after we downloaded it, if not, no need to include it in updates
+		downloadedFileDetails, ok := fc.files[uploader]
+		if !ok || downloadedFileDetails.mtime != object.ModifiedAt {
+			toDownload = append(toDownload, archivedFileDetails{uploader: uploader, mtime: object.ModifiedAt})
+		}
+	}
+
+	for uploader, fileDetails := range fc.files {
+		if _, isOK := listedUploaders[uploader]; !isOK {
+			toDelete = append(toDelete, archivedFileDetails{uploader: uploader, mtime: fileDetails.mtime})
+		}
+	}
+
+	return
+}
+
+// downloadFile downloads a file from archive to cache.
+// It first downloads it to a temp file so that we close the existing file(if already exists), replace it with new one and then reopen it.
+func (a *Archiver) downloadFile(ctx context.Context, period string, afd archivedFileDetails, fc *filesCollection) error {
+	objectKey := fmt.Sprintf("%s/%s", period, afd.uploader)
+	readCloser, err := a.objectClient.GetObject(ctx, objectKey)
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		if err := readCloser.Close(); err != nil {
+			level.Error(util.Logger)
+		}
+	}()
+
+	downloadPath := path.Join(a.cfg.CacheLocation, period)
+	err = chunk_util.EnsureDirectory(downloadPath)
+	if err != nil {
+		return err
+	}
+
+	// download the file temporarily with some other name to allow boltdb client to close the existing file first if it exists
+	tempFilePath := path.Join(downloadPath, fmt.Sprintf("%s.%d", afd.uploader, time.Now().Unix()))
+
+	f, err := os.Create(tempFilePath)
+	if err != nil {
+		return err
+	}
+
+	_, err = io.Copy(f, readCloser)
+	if err != nil {
+		return err
+	}
+
+	fc.Lock()
+	defer fc.Unlock()
+
+	af, ok := fc.files[afd.uploader]
+	if ok {
+		if err := af.boltdb.Close(); err != nil {
+			return err
+		}
+	} else {
+		af = archivedFile{downloadLocation: path.Join(downloadPath, afd.uploader)}
+	}
+
+	// move the file from temp location to actual location
+	err = os.Rename(tempFilePath, af.downloadLocation)
+	if err != nil {
+		return err
+	}
+
+	af.mtime = afd.mtime
+	af.boltdb, err = local.OpenBoltdbFile(af.downloadLocation)
+	if err != nil {
+		return err
+	}
+
+	fc.files[afd.uploader] = af
+
+	return nil
+}
+
+// uploadFile uploads one of the files locally written by ingesters to archive.
+func (a *Archiver) uploadFile(ctx context.Context, period string) error {
+	if a.cfg.Mode == ArchiverModeReadWrite {
+		return nil
+	}
+
+	snapshotPath := path.Join(a.cfg.CacheLocation, period)
+	err := chunk_util.EnsureDirectory(snapshotPath)
+	if err != nil {
+		return err
+	}
+
+	filePath := path.Join(snapshotPath, fmt.Sprintf("%s.%d", a.uploader, time.Now().Unix()))
+	f, err := os.Open(filePath)
+	if err != nil {
+		return err
+	}
+
+	db, err := a.localBoltdbGetter(period, local.DBOperationRead)
```

Comment:
I think ingester locally should also use the localdb for the store when you query the store for extra data that might not have been delivered yet to others. It looks like currently if we are a readwriter (single binary) archiver ingester will also return data from other ingester, this is not necessary and will generates a lot of duplicates.

see https://github.com/grafana/loki/pull/1837#discussion_r397408387 WDYT

Reply:
I can't think of cleaner ways to limit ingesters from querying data which only it has generated in single binary mode. Do you have any thoughts?